### PR TITLE
Add ensure capacity for enumerable types.

### PR DESCRIPTION
**Pull request by TimothyMakkison**
Complete #251 for enumerable types. I'm not sure how Explicit `Count` values should be treated, right now it gets the runtime type using the `is` operator. I think it should either try not to get the value or just cast the value so the explicit property can be accessed.

- Added `EnsureCapacityBuilder`, `CanEnsureCapacity` checks if ensure capacity can be used, storing the relevant information in `EnsureCapacityInfo` to be used by `TryBuildEnsureCapacityStatement` when the syntax is built.
  - If both source and target have compile time sizes `TryBuildEnsureCapacityStatement` generates a simple call to `EnsureCapacity`.
  - If source does not have a length/count property, `is` statements are created to check the runtime type, invoking `EnsureCapacity` with the resulting type.
  - Not sure where `EnsureCapacityBuilder` should be placed.
- Added tests, waiting for integration tests from github actions.

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Riok.Mapperly.Helpers;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static Riok.Mapperly.Emit.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+
+public class EnsureCapacityInfo
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:11]: This code base usually follows one type per file approach.

To match existing architecture convert these properties to private fields and create an exposed method `BuildEnsureCapacity()` which creates the syntax (content of `TryBuildEnsureCapacityStatement`) (similar to how the object factories and mappings work).

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    private static bool TryGetNonEnumeratedCount(ITypeSymbol value, WellKnownTypes types, [NotNullWhen(true)] out string? expression)
+    {
+        if (value.IsArrayType())
+        {
+            expression = "Length";
+            return true;
+        }
+        if (value.HasImplicitInterfaceProperty(types.ICollectionT, "Count") || value.HasImplicitInterfaceProperty(types.IReadOnlyCollectionT, "Count"))
+        {
+            expression = "Count";
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:124]: Introduce const's and probably nameofs for these strings.

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            expression = EnsureCapacityStatement(target, MemberAccess(ctx.Source, state.SourceAccessor), MemberAccess(target, state.TargetAccessor));
+            return true;
+        }
+
+        // generate check for runtime type, calling EnsureCapacity if it is a collection/readonlyCollection
+        // this is used if source is IEnumerable
+        var targetCount = MemberAccess(target, state.TargetAccessor);
+
+        var ifCollection = IfIsTypeEnsureCapacityStatement("collection", state.CollectionType!, ctx, target, targetCount);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:89]: If you use named identifiers in the generated code always make sure to use `ctx.NameBuilder.New` to ensure that the name is unique.

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+        // if source accessor is avaailable then create simple EnsureCapacity call
+        if (state.SourceAccessor is not null)
+        {
+            expression = EnsureCapacityStatement(target, MemberAccess(ctx.Source, state.SourceAccessor), MemberAccess(target, state.TargetAccessor));
+            return true;
+        }
+
+        // generate check for runtime type, calling EnsureCapacity if it is a collection/readonlyCollection
+        // this is used if source is IEnumerable
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:86]: If the target framework supports `Enumerable.TryGetNonEnumeratedCount`, I'd use that instead. This can be found out if the `TryGetNonEnumeratedCount` method symbol is available on the enumerable type.
Probably use a type switch / pattern matching as fallback.

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static Riok.Mapperly.Emit.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+
+public class EnsureCapacityInfo
+{
+    public string? SourceAccessor { get; set; }
+    public string TargetAccessor { get; set; } = null!;
+    public TypeSyntax? CollectionType { get; set; }
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:15]: I'd stick to the symbols here instead of using the syntax and just build the typesyntax in the build method, if it is required there.

File: test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs [hunk 6]
```diff
@@ -565,6 +601,52 @@ public void EnumerableToCreatedQueue()
...
+            "A",
+            "B",
+            "class A { public int[] Value { get; } }",
+            "class B { public Stack<long> Value { get; set; } }");
+        TestHelper.GenerateMapper(source)
+            .Should()
+            .HaveMethodBody("MapToStack",
+                """
+                var target = new System.Collections.Generic.Stack<long>();
+                target.EnsureCapacity(source.Length + target.Count);
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:616]: This is not yet optimal, the built in collections all have a ctor accepting a capacity parameter which would reduce the array allocations and could be used by Mapperly. However, this could also be adressed in a separate PR.

**Review by latonz — CHANGES_REQUESTED** [review]
Thank you for another great contribution 🎉 

I'd create a new namespace `Riok.Mapperly.Descriptors.Enumerables` for these two classes (`EnsureCapacityInfo` and `EnsureCapacityBuilder`)

File: src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs (hunk 4 above)
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:86]: I'd thought of doing that but wasn't sure how to check which framework is being used. I could check if Enumerable has the corresponding method.

File: test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs (hunk 6 above)
Comment (TimothyMakkison) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:616]: > However, this could also be adressed in a separate PR.

I'd planned on updating the issue after completing this PR. I was going to look at how you did this with dictionaries.

What concrete types is this needed for? Types like `Stack<T>`, `Queue<T>` have constructors that take `IEnumerable<T>`  so I was going to use that where possible.

Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:616]: Sounds good to me. `Stack<T>`, `Queue<T>` are all, as I think `List<T>` and co are always mapped to via linq.

Comment (TimothyMakkison) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:616]: Should I make those changes a separate PR?

Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:616]: I'd do it separately if possible.

**Comment by TimothyMakkison** [comment]
- Added version specific `TryGetNonEnumeratedCount` - I've updated the integration test to test this.
- Only use symbols until ensure capacity is built.
- Refactored `EnsureCapacityBuilder` into 3 different versions - wasnt a fan of creating 5 files but it felt better than using ad-hoc discriminant union. I can convert it back to the previous version if you prefer.
- All generated code uses the global type.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilderIsType.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    public EnsureCapacityBuilderIsType(string targetAccessor, INamedTypeSymbol collectionType, INamedTypeSymbol readonlyCollectionType)
+    {
+        _targetAccessor = targetAccessor;
+        _collectionType = collectionType;
+        _readonlyCollectionType = readonlyCollectionType;
+    }
+
+    public StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target)
+    {
+        var targetCount = MemberAccess(target, _targetAccessor);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilderIsType.cs:28]: To improve the readability add comments with how the generated code looks like.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityHelper.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            expression = CountMethodName;
+            return true;
+        }
+
+        expression = null;
+        return false;
+    }
+
+    public static ExpressionStatementSyntax EnsureCapacityStatement(ExpressionSyntax target, ExpressionSyntax sourceCount, ExpressionSyntax targetCount)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityHelper.cs:78]: I'd move this syntax conversion into an abstract base class EnsureCapacity to follow the concept of MappingBuilders and Mappings.

File: src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs [hunk 9]
```diff
@@ -74,18 +75,24 @@ public static class EnumerableMappingBuilder
...
         // create a foreach loop with add calls if source is not an array
         // and  ICollection.Add(T): void is implemented and not explicit
         // ensures add is not called and immutable types
         if (!ctx.Target.IsArrayType() && ctx.Target.HasImplicitInterfaceMethod(ctx.Types.ICollectionT, AddValueMethodName))
-            return new ForEachAddEnumerableExistingTargetMapping(ctx.Source, ctx.Target, elementMapping, AddValueMethodName);
+            return CreateForEach(AddValueMethodName);
 
         return null;
+
+        ForEachAddEnumerableExistingTargetMapping CreateForEach(string propertyName)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:91]: Is the parameter name `propertyName` correct here?

File: src/Riok.Mapperly/Helpers/SymbolExtensions.cs [hunk 10]
```diff
@@ -103,6 +103,29 @@ internal static bool HasImplicitInterfaceMethod(this ITypeSymbol symbol, INamedT
         return !methodInterImplementaton.ExplicitInterfaceImplementations.Any();
     }
 
+    internal static bool HasImplicitInterfaceProperty(this ITypeSymbol symbol, INamedTypeSymbol inter, string propertyName)
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/SymbolExtensions.cs:106]: Do we need the limitation of properties here?
I'd probably prefer the approach of https://github.com/riok/mapperly/pull/341#discussion_r1167471503.

File: test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs [hunk 11]
```diff
@@ -22,7 +22,7 @@ public TestObjectDto(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
         public int IntInitOnlyValue { get; init; }
 
 #if NET7_0_OR_GREATER
-        public required int RequiredValue { get; init; }
+        public int RequiredValue { get; init; }
```
Comment (latonz) [inline test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:25]: Mhmmm, is this change expected?

File: test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs [hunk 12]
```diff
@@ -21,7 +21,9 @@ public TestObject(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
         public int IntInitOnlyValue { get; init; }
 
 #if NET7_0_OR_GREATER
-        public required int RequiredValue { get; init; }
+
+        public int RequiredValue { get; init; }
```
Comment (latonz) [inline test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs:25]: Mhmmm, is this change expected?

File: test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs [hunk 13]
```diff
@@ -53,6 +53,8 @@ public TestObjectDto(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 
         public TestObjectNestedDto[]? NullableReadOnlyObjectCollection { get; set; }
 
+        public Stack<int> EnsureCapacityValue { get; set; } = new();
```
Comment (latonz) [inline test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:56]: Is this separate test property really required? What additional test case does it test?

**Review by latonz — CHANGES_REQUESTED** [review]
I like the new approach. 
I'd adjust the names to the general concept of Mapperly with MappingBuilders and Mappings:
> The idea is to use a MappingBuilder to construct a Mapping object. The Mapping then should only transform the given information into a syntax. The MappingBuilder should provide all the required information for the Mapping to do only the transformation.

I'd rename the `EnsureCapacityBuilder*` to `EnsureCapacity*` (as they are not builder in the sense of Mapperly). However, the `EnsureCapacityHelper` would be a builder.
`IEnusreCapacityBuilder` => `EnsureCapacity` (abstract base class)
`EnsureCapacityHelper` => `EnsureCapacityBuilder`
`EnsureCapacityBuilderNonEnumerated` => `EnsureCapacityNonEnumerated`
`EnsureCapacityBuilderSimple` => `EnsureCapacityMember`
`EnsureCapacityBuilderIsType` => `EnsureCapacityWellKnownTypes`

File: test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs (hunk 13 above)
Comment (TimothyMakkison) [reply test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:56]: `Enumerable.TryGetNonEnumeratedCount` was added in .Net 6+. If a lower version is used a statement using pattern matching is created (`EnsureCapacityBuilderIsType`). I took advantage of the .Net Framework 4.8 integration test to ensure that this works.

I could probably remove `EnsureCapacityBuilderIsType` as its never used in .Net 6+

File: test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs (hunk 12 above)
Comment (TimothyMakkison) [reply test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs:25]: mb I used Codemaid for linting and it still doesn't support modern C# 😞

File: test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs (hunk 13 above)
Comment (latonz) [reply test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:56]: Ah now I got it, this tests a mapping `IEnumerable<string>` => `Stack<int>` (which uses the type pattern matching / TryGetNonEnumeratedCount respectively), `Stack<string>` => `Stack<int>` tests the case for which the source count is available directly.
I'd keep `EnsureCapacityIsType` for now, as Mapperly tries to support .NET Framework as good as possible.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public abstract class EnsureCapacity
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+
+    public abstract StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target);
+
+    public static ExpressionStatementSyntax EnsureCapacityStatement(ExpressionSyntax target, ExpressionSyntax sourceCount, ExpressionSyntax targetCount)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs:16]: This method should be protected, shouldn't it?

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        value.ImplementsGeneric(types.ICollectionT, CountMethodName, out var _, out var hasCollectionCount);
+        if (hasCollectionCount)
+        {
+            expression = CountMethodName;
+            return true;
+        }
+
+        value.ImplementsGeneric(types.IReadOnlyCollectionT, CountMethodName, out var _, out var hasReadOnlyCount);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:71]: Nit: the `var` keyword in `out var _` is not needed, is it? (same on L64).

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+    private const string CountMethodName = nameof(ICollection<object>.Count);
+    private const string LengthMethodName = nameof(Array.Length);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:11]: These are not really method but property names...

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
+
+    public static EnsureCapacity? TryCreateEnsureCapacityBuilder(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:14]: I'd simplify this method's name (it doesn't return a builder). Eg. `TryBuildEnsureCapacity` (would also align with `TryBuildMapping`).

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+using static Riok.Mapperly.Emit.SyntaxFactoryHelper;
+
+
+namespace Riok.Mapperly.Descriptors.Enumerables.EnsureCapacity;
+
+public abstract class EnsureCapacity
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+
+    public abstract StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs:14]: I'd simplify this method name. There is only one build method in this class which is exposed and since the class is named `EnsureCapacity` it should be clear that it is an ensure capacity statement. IMO `Build` should be simple and understandable.

File: src/Riok.Mapperly/Helpers/SymbolExtensions.cs [hunk 19]
```diff
@@ -79,6 +79,49 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
...
+
+        genericIntf = t.AllInterfaces.FirstOrDefault(x => x.IsGenericType && SymbolEqualityComparer.Default.Equals(x.OriginalDefinition, genericInterfaceSymbol));
+
+        if (genericIntf == null)
+        {
+            isExplicit = false;
+            return false;
+        }
+
+        var interfaceSymbol = genericIntf.GetMembers(symbolName).Single();
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/SymbolExtensions.cs:104]: TBD: maybe switching to `First` instead of `Single` will be more resilient if a future release adds a method overload for a given method.

File: test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs [hunk 20]
```diff
@@ -22,6 +22,7 @@ public TestObject(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 
 #if NET7_0_OR_GREATER
         public required int RequiredValue { get; init; }
+
```
Comment (latonz) [inline test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs:25]: Nit: Probably revert this change?

File: test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs [hunk 21]
```diff
@@ -631,6 +631,47 @@ public void MapToExistingExplicitAddCustomCollection()
             .HaveSingleMethodBody("");
     }
 
+    [Fact]
+    public void MapToExistingExplicitCountCustomCollection()
+    {
+        // should not create an EnsureCapacity statement when the target has an explicit count
+        var source = TestSourceBuilder.MapperWithBodyAndTypes(
+            "partial void Map(IEnumerable<int> source, B target);",
+            "class B : List<int> { int List<int>.Count { get; } }");
```
Comment (latonz) [inline test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:640]: I don't get this test case, class B extends the List class, but implements the base class property explicit. This does not really work, does it?

File: test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs [hunk 22]
```diff
@@ -319,7 +323,18 @@ public static partial class StaticTestMapper
             return target;
         }
 
-        private static global::System.Collections.Generic.Stack<int> MapToStack(global::System.Collections.Generic.Stack<string> source)
+        private static global::System.Collections.Generic.Stack<int> MapToStack(global::System.Collections.Generic.IEnumerable<string> source)
+        {
+            var target = new global::System.Collections.Generic.Stack<int>();
+            foreach (var item in source)
```
Comment (latonz) [inline test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs:329]: Shouldn't this be the point where the type tests should be emitted? As mentioned in another comment, I'm also ok with removing the `EnsureCapacityIsType` but then this integration test property could also be removed.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+        // if target does not have a count then return null
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+            return null;
+
+        // if target and source count are known then create a simple EnsureCapacity statement
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:34]: Instead of asserting `iEnumerable` is non-null later, I'd if + return null here:
```c#
if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
    return null;
```

**Review by latonz — CHANGES_REQUESTED** [review]
Looks pretty good 😊 Thank you a lot for your efforts!

A few nitpicks left and then this should be ready to merge 🚀

File: src/Riok.Mapperly/Helpers/SymbolExtensions.cs [hunk 24]
```diff
@@ -79,6 +79,49 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
...
+        {
+            IMethodSymbol methodSymbol => !methodSymbol.ExplicitInterfaceImplementations.Any(),
+            IPropertySymbol propertySymbol => !propertySymbol.ExplicitInterfaceImplementations.Any(),
+            _ => throw new NotImplementedException(),
+        };
+
+        return true;
+    }
+
     internal static bool HasImplicitInterfaceMethod(this ITypeSymbol symbol, INamedTypeSymbol inter, string methodName)
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/SymbolExtensions.cs:125]: This could now be simplified a lot, couldn't it? (eg. `symbol.ImplementsGeneric(inter, methodName, out var isExplicit) && !isExplicit`

File: test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs (hunk 21 above)
Comment (TimothyMakkison) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:640]: Yeah this test is terrible for many reasons. I decided not to generate an EnsureCapacity statement when the target type has an `EnsureCapacity` method and an explicit length. This is because I deemed it to be an edge case and a complete pain to handle with `EnsureCapacityIsType`. 

For some reasons I made a (broken) test to ensure it won't generate an `EnsureCapacity` statement under these conditions.

I could either delete this test, fix it, or update the logic to insert a cast to cast the target when its `Count` is explicit.

Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:640]: I'm okay with deleting this test.

**Comment by TimothyMakkison** [comment]
Turns out `EnsureCapacity` wasn't added until [.Net 6+](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.ensurecapacity?view=net-8.0#system-collections-generic-list-1-ensurecapacity(system-int32)), so `EnsureCapacityIsType` and the integration test are not needed.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+        // if target does not have a count then return null
+        if (!TryGetNonEnumeratedCount(targetType, types, out var targetSizeProperty))
+            return null;
+
+        // if target and source count are known then create a simple EnsureCapacity statement
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
+            return null;
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:35]: Is this ever reached? `TryBuildEnsureCapacity` is only called for enumerables, isn't it?

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+
+        var nonEnumeratedCountMethod = types.Enumerable.GetMembers(TryGetNonEnumeratedCountMethodName)
+                                                        .OfType<IMethodSymbol>()
+                                                        .FirstOrDefault(x => x.ReturnType.SpecialType == SpecialType.System_Boolean && x.IsStatic && x.Parameters.Length == 2 && x.IsGenericMethod);
+
+        // if source does not have a count use GetNonEnumeratedCount, calling EnusureCapacity if count is available
+        var typedNonEnumeratedCount = nonEnumeratedCountMethod.Construct(iEnumerable.TypeArguments.ToArray());
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:42]: This leads to a null ref exception on .NET Framework... add a null guard:
```c#
if (nonEnumeratedCountMethod == null)
    return null;

...
```

**Review by latonz — CHANGES_REQUESTED** [review]

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs (hunk 26 above)
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:42]: .NET Framework can't get to this section, [line 23](https://github.com/riok/mapperly/pull/312/files/f19efb7c50ed837ab5986cf93b4ac58f64072d96#diff-3f8932b41774a45b65c962c4ae183d71993fe8112d68981d3ab20f2a1d53172fR23) prevents non .NET 6+ code from getting here because they don't have `EnsureCapacity` available yet.

File: src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs (hunk 25 above)
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:35]: I used ImplementsGeneric to get the typed version of `IEnumerableT`. I originally didn't use this redundant null, but changed it when you [suggested I remove the null suppression](https://github.com/riok/mapperly/pull/312#discussion_r1167799422)
